app/javascript/extensible-components/lib.ts

```diff
@@ -0,0 +1,26 @@
+export interface IExtensionComponent {
+  action: string;
+  payload: any;
+}
+
+export interface IMiQApiCallback {
+  [propName: string]: Function;
+}
+
+export interface IExtensibleComponent {
+  extensibleComponent: any;
+  apiCallbacks: () => IMiQApiCallback;
+  renderCallbacks: () => IMiQApiCallback;
+}
+
+export const getItems = ManageIQ.extensions.getItems;
+
+/**
+ * Helper function to create new component.
+ * @param name string name of new component.
+ * @param api callback functions to change inner logic of component.
+ * @param render callback function to apply render functions.
+ */
+export function addComponent(name: string, api?: IMiQApiCallback, render?: IMiQApiCallback) {
```

Comment:
Hmm, we could also do:

```
export const addComponent = ManageIQ.extensions.addComponent;
```

to cut down the boilerplate code. But I'm not sure how TypeScript fits the above syntax form 😕

Reply:
We can, however I am not fan of assigning functions to const. So to be consistent we should all three things export as functions `getItems` `subscribe` `addComponent`

Reply:
@karelhala just wondering - why you are not a fan?

Reply:
@ohadlevy well mostly due to readability. I really like fat arrows, but for named functions I prefer the "old" way instead of assign to const :smile: 
```typescript
function getTen() {
   return 10;
}
```
Is somewhat more readable and from first glance you see that it's function, instead when reading this
```Typescript
const getTen = () => 10;
```
It looks nice and all, but the person reading this has to know that it's function assignment into const.

Plus binding is kinda strange with the second one.
```Typescript
function printA() {
  console.log(this.a);
}

const printA2 = () => console.log(this.a)

printA.bind({a: 'someText'})();
printA2.bind({a: 'someText'})();
```
The second one ignores binding of object, I know what it's the reason how `this` is handled with `let` and `const`, but that's not how named functions should behave. Again for anonymous functions it's the best things to do (no need to bind this to every function).

Reply:
I agree that lambdas declared through `=>` are best used for anonymous functions.

From variable scoping point of view, `const foo = function() {...}` should be the same as `function foo() {...} `. Also, from TypeScript perspective:

```
const add1 = (a: number, b: number): number => { return a + b; };
function add2 (a: number, b: number): number { return a + b; }
```

Reply:
> Right, lambdas don't have their own this and this should be taken into account when passing those around.

Exactly, but
```javascript
const getTen = () => 10;

typeof(getTen);
```
Will return `function` so one could assume it's "normal" function and `call`, `bind`, `apply` and such will behave differently (of course programmer would not call `this` inside lambda expecting to be set by someone else). But you get the picture why I prefer plain old `function` for named functions instead of assign to `const`